proxy/pkg/zdmproxy/clientconn.go

```diff
@@ -174,14 +175,15 @@ func (cc *ClientConnector) listenForRequests() {
 					err, cc.clientHandlerContext, cc.clientHandlerCancelFunc, ClientConnectorLogPrefix, "reading", connectionAddr)
 				break
 			} else if protocolErrResponseFrame != nil {
-				f = protocolErrResponseFrame
-				if !protocolErrOccurred {
-					protocolErrOccurred = true
-					cc.sendResponseToClient(protocolErrResponseFrame)
-					cc.clientHandlerShutdownRequestCancelFn()
-					setDrainModeNowFunc()
-					continue
-				}
+				alreadySentProtocolErr = protocolErrResponseFrame
+				protocolErrOccurred = true
+				cc.sendResponseToClient(protocolErrResponseFrame)
+				continue
+			} else if alreadySentProtocolErr != nil {
```

Comment:
Just to check that I understand correctly. The intended behaviour with this change is: if a protocol error is sent on the connection, then we want to ensure that all other potentially pending responses that may still be returned are replaced by that protocol error (unless they were themselves protocol errors in the first place). So we're keeping the already returned protocol error to replace the frame of any subsequent non-protocol error response with it.

The idea is to ensure that we do not send seemingly valid responses on a connection on which a protocol error was returned while this connection is being drained. 

Is this correct?

Reply:
Yes, but also the main change is that we no longer trigger the connection shutdown mechanism so the connection will be left in an unusable state (constant protocol errors) but it will be up to the driver to close it (at least until we address #68 )